File: lang/ruby/test_identity.rb
# frozen_string_literal: true
#
# ID/IDREF tests for the Ruby CX binding (ADR 0003).
# Mirrors V conformance/identity.txt and lang/python/test_identity.py.
# Run: ruby lang/ruby/test_identity.rb
#
require_relative 'cxlib/lib/cxlib'

$passed = 0
$failed = 0

def run_test(name)
  yield
  $passed += 1
  print "  PASS  #{name}\n"
rescue => e
  $failed += 1
  print "  FAIL  #{name}\n"
  print "    #{e.class}: #{e.message}\n"
  print "    #{e.backtrace.first(2).join("\n    ")}\n"
end

def assert_eq(expected, actual, msg = nil)
  raise "expected #{expected.inspect}, got #{actual.inspect}#{" (#{msg})" if msg}" \
    if expected != actual
end

def assert_true(cond, msg = nil)
  raise "expected true#{" (#{msg})" if msg}" unless cond
end

def root(d)
  d.elements.find { |e| e.is_a?(CXLib::Element) }
end

run_test('id_declaration_only_round_trips') do
  cx_in = '[user #u-1 name=alice]'
  doc = CXLib.parse(cx_in)
  assert_eq 'u-1', root(doc).id
  assert_eq cx_in, doc.to_cx
end

run_test('id_with_anchor_coexists') do
  doc = CXLib.parse('[item &a #u-1 v=42]')
  item = root(doc)
  assert_eq 'a',   item.anchor
  assert_eq 'u-1', item.id
end

run_test('attribute_value_reference_marked_is_ref') do
  doc = CXLib.parse('[users [user #u-1 name=alice] [reviewer assigned-to=@u-1]]')
  reviewer = doc.find_first('reviewer')
  a = reviewer.attrs.find { |x| x.name == 'assigned-to' }
  assert_true a.is_ref
  assert_eq 'u-1', a.value
end

run_test('resolve_id_finds_declared_element') do
  doc = CXLib.parse('[users [user #u-1 name=alice] [user #u-2 name=bob]]')
  assert_eq 'alice', doc.resolve_id('u-1').attr('name')
  assert_eq 'bob',   doc.resolve_id('u-2').attr('name')
  assert_true doc.resolve_id('u-3').nil?
end

run_test('elements_by_id_builds_full_map') do
  doc = CXLib.parse('[a #x v=1] [b #y v=2] [c #z v=3]')
  m = doc.elements_by_id
  assert_eq 3, m.size
  assert_eq 'a', m['x'].name
  assert_eq 'b', m['y'].name
  assert_eq 'c', m['z'].name
end

run_test('quoted_at_literal_is_not_a_reference') do
  cx_in = "[item label='@literal']"
  doc = CXLib.parse(cx_in)
  label = root(doc).attrs.find { |a| a.name == 'label' }
  assert_true !label.is_ref
  assert_eq '@literal', label.value
  assert_eq cx_in, doc.to_cx
end

run_test('forward_reference_resolves') do
  doc = CXLib.parse('[users [reviewer assigned-to=@u-1] [user #u-1 name=alice]]')
  user = doc.resolve_id('u-1')
  assert_true !user.nil?
  assert_eq 'alice', user.attr('name')
end

run_test('nested_id_and_ref_round_trip') do
  cx_in = "[doc\n  [users\n    [user #u-1 name=alice]\n  ]\n  [reviews\n    [review target=@u-1 score=5]\n  ]\n]"
  doc = CXLib.parse(cx_in)
  assert_true !doc.resolve_id('u-1').nil?
  review = doc.find_first('review')
  target = review.attrs.find { |a| a.name == 'target' }
  assert_true target.is_ref
  assert_eq 'u-1', target.value
end

run_test('body_ref_survives_ast_bin_round_trip') do
  # Phase 7.70 — ast_bin v3 carries body_ref through the V↔binding
  # boundary. The field is populated post-parse from the v3 wire bytes,
  # not re-detected from text.
  cx_in = '[doc [section #section-3 [para See [ref @section-3].]]]'
  doc = CXLib.parse(cx_in)
  section = doc.find_first('section')
  para = section.find_first('para')
  body_ref_node = para.items.find { |c| c.is_a?(CXLib::Element) && c.name == 'ref' }
  assert_true !body_ref_node.nil?
  assert_eq 'section-3', body_ref_node.body_ref
  assert_eq [], body_ref_node.attrs
  assert_eq [], body_ref_node.items
  assert_true doc.to_cx.include?('[ref @section-3]')
end

run_test('multiple_refs_to_same_id') do
  doc = CXLib.parse(
    "[users [user #u-1 name=alice] " \
    "[reviewer assigned-to=@u-1] " \
    "[approver checked-by=@u-1]]"
  )
  count = 0
  (doc.find_all('reviewer') + doc.find_all('approver')).each do |el|
    el.attrs.each { |a| count += 1 if a.is_ref && a.value == 'u-1' }
  end
  assert_eq 2, count
end

print "\n#{$passed}/#{$passed + $failed} passed\n"
exit($failed == 0 ? 0 : 1)

File: lang/python/test_identity.py
"""Tests for ID/IDREF resolution per spec/identity.md.

Mirrors the V core conformance/identity.txt cases for the Python binding
accessor surface (`Element.id`, `Attr.is_ref`, `Document.resolve_id()`,
`Document.elements_by_id()`). The fields are populated by the V core
parser and threaded through the v2 ast_bin wire format; the binding
re-emits via `to_cx()` so we also check the round-trip.
"""
from cxlib import parse
from cxlib.ast import Element


def _root(doc):
    return doc.root()


def test_id_declaration_only_round_trips():
    cx_in = "[user #u-1 name=alice]"
    doc = parse(cx_in)
    user = _root(doc)
    assert user.id == "u-1"
    assert doc.to_cx() == cx_in


def test_id_with_anchor_coexists():
    cx_in = "[item &a #u-1 v=42]"
    doc = parse(cx_in)
    item = _root(doc)
    assert item.anchor == "a"
    assert item.id == "u-1"
    assert doc.to_cx() == cx_in


def test_attribute_value_reference_marked_is_ref():
    doc = parse("[users [user #u-1 name=alice] [reviewer assigned-to=@u-1]]")
    reviewer = doc.find_first("reviewer")
    ref_attr = next(a for a in reviewer.attrs if a.name == "assigned-to")
    assert ref_attr.is_ref is True
    assert ref_attr.value == "u-1"


def test_resolve_id_finds_declared_element():
    doc = parse("[users [user #u-1 name=alice] [user #u-2 name=bob]]")
    el = doc.resolve_id("u-1")
    assert el is not None
    assert el.attr("name") == "alice"
    assert doc.resolve_id("u-2").attr("name") == "bob"
    assert doc.resolve_id("u-3") is None


def test_elements_by_id_builds_full_map():
    doc = parse("[a #x v=1] [b #y v=2] [c #z v=3]")
    by_id = doc.elements_by_id()
    assert set(by_id.keys()) == {"x", "y", "z"}
    assert by_id["x"].name == "a"
    assert by_id["y"].name == "b"
    assert by_id["z"].name == "c"


def test_quoted_at_literal_is_not_a_reference():
    cx_in = "[item label='@literal']"
    doc = parse(cx_in)
    item = _root(doc)
    label = next(a for a in item.attrs if a.name == "label")
    assert label.is_ref is False
    assert label.value == "@literal"
    assert doc.to_cx() == cx_in


def test_forward_reference_resolves():
    doc = parse(
        "[users [reviewer assigned-to=@u-1] [user #u-1 name=alice]]"
    )
    user = doc.resolve_id("u-1")
    assert user is not None
    assert user.attr("name") == "alice"


def test_nested_id_and_ref_round_trip():
    cx_in = "[doc\n  [users\n    [user #u-1 name=alice]\n  ]\n  [reviews\n    [review target=@u-1 score=5]\n  ]\n]"
    doc = parse(cx_in)
    user = doc.resolve_id("u-1")
    assert user is not None
    review = doc.find_first("review")
    target = next(a for a in review.attrs if a.name == "target")
    assert target.is_ref is True
    assert target.value == "u-1"


def test_multiple_refs_to_same_id():
    doc = parse(
        "[users [user #u-1 name=alice] "
        "[reviewer assigned-to=@u-1] "
        "[approver checked-by=@u-1]]"
    )
    refs = [a for el in doc.find_all("reviewer") + doc.find_all("approver")
            for a in el.attrs if a.is_ref]
    assert len(refs) == 2
    assert all(r.value == "u-1" for r in refs)
    assert doc.resolve_id("u-1").attr("name") == "alice"


def test_body_ref_survives_ast_bin_round_trip():
    """Phase 7.70 — ast_bin v3 carries body_ref through the V↔binding
    boundary. The field is populated post-parse from the v3 wire bytes,
    not re-detected from text."""
    cx_in = "[doc [section #section-3 [para See [ref @section-3].]]]"
    doc = parse(cx_in)
    section = doc.find_first("section")
    para = section.find_first("para")
    body_ref_node = next(c for c in para.items
                          if isinstance(c, Element) and c.name == "ref")
    assert body_ref_node.body_ref == "section-3"
    assert body_ref_node.attrs == []
    assert body_ref_node.items == []
    # body_ref participates in the binding's local CX-text emit
    assert "[ref @section-3]" in doc.to_cx()


def main():
    import sys, traceback
    tests = [g for n, g in sorted(globals().items()) if n.startswith("test_") and callable(g)]
    passed = 0
    failed = 0
    for t in tests:
        try:
            t()
            print(f"ok  {t.__name__}")
            passed += 1
        except Exception:
            traceback.print_exc()
            print(f"FAIL {t.__name__}")
            failed += 1
    print(f"\n{passed}/{passed + failed} passed")
    sys.exit(1 if failed else 0)


if __name__ == "__main__":
    main()
